macros/contexts/contextNondecimalBase.pl

```diff
@@ -0,0 +1,273 @@
+
+package context::NondecimalBase;
+
+# defines the base for the context.
+our $base = 10;
+#
+#  The standard digits, pre-built so it doesn't have to be done each time the conversion is called
+#
+our $digits16 = [ '0' .. '9', 'A' .. 'F' ];
+our $digit16  = { map { ($digits16->[$_], $_) } (0 .. scalar(@$digits16) - 1) };
+
+use Data::Dumper;
+#
+#  Initialize the contexts and make the creator function.
+#
+sub Init {
+	my $context = $main::context{NondecimalBase} = Parser::Context->getCopy("Numeric");
+	$context->{name} = 'NondecimalBase';
+	$context->{parser}{Number} = 'context::NondecimalBase::Number';
+	# $context->{value}{Real} = 'context::Hex::Hex';
+	$context->{pattern}{number} = '[0-9A-F]+';
+	$context->functions->disable('All');
+	$context->operators->remove('^');
+	$context->parens->remove('|');
+	$context->constants->clear();
+	# $context->operators->add(
+	#   '&'  => {precedence => .6, associativity => 'left', type => 'bin', string => ' & ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] & $_[1]}},
+	#   '|'  => {precedence => .5, associativity => 'left', type => 'bin', string => ' | ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] | $_[1]}},
+	#   '^'  => {precedence => .5, associativity => 'left', type => 'bin', string => ' ^ ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] ^ $_[1]}},
+	#   '>>' => {precedence => .4, associativity => 'left', type => 'bin', string => ' >> ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] >> $_[1]}},
+	#   '<<' => {precedence => .4, associativity => 'left', type => 'bin', string => ' << ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] << $_[1]}},
+	#   '~'  => {precedence => 6, associativity => 'left', type => 'unary', string => '~',
+	#            class => 'context::Hex::UOP::not'},
+	# );
+	# $context->{precedence}{Hex} = $context->{precedence}{special};
+	$context->flags->set(limits => [ -1000, 1000, 1 ]);
+	$context->update;
+
+	# main::PG_restricted_eval('sub Hex {context::Hex::Hex->new(@_)}');
+}
+use Data::Dumper;
+
+sub convert {
+	my $value = shift;
+	print Dumper "convert: $value";
+	# Set default options and get passed in options.
+	my %options = (
+		from   => 10,
+		to     => 10,
+		digits => $digits16,
+		@_
+	);
+	my $from   = $options{'from'};
+	my $to     = $options{'to'};
+	my $digits = $options{'digits'};
+	print Dumper "from: $from; to: $to";
+
+	die "The digits option must be an array of characters to use for the digits"
+		unless ref($digits) eq 'ARRAY';
+
+	#
+	# The highest base the digits will support
+	#
+	my $maxBase = scalar(@$digits16);
+
+	die "The base of conversion must be between 2 and $maxBase"
+		unless $to >= 2 && $to <= $maxBase && $from >= 2 && $from <= $maxBase;
+
+	#
+	# Reverse map the digits to base 10 values
+	#
+	my $baseBdigits = { map { ($digits->[$_], $_) } (0 .. $from - 1) };
```

Comment:
While this works, it does mean you are making this array over and over again as you convert, even when that isn't necessary in most cases.  If the user doesn't specify the `digits` option, you can use the `$digit16` mapping without having to rebuild it each time.  That is the point of `$digit16`.  The code I provided includes a test here for whether `$digits == $digits16` and uses `$digit16` in that case for efficiency.  I recommend you do that here.

In fact, since you are now using a context, you could have the digits be a context flag as well, and use `setBase()` to set both the base and the allowed digits.  Then you can create the reverse mapping in `setBase()` and only do it once, even if alternative digits are specified.  You could then make `convert()` take a context as an argument so that you can read the digits from it (or use them as the default digits) and the reverse map.  It is also possible to subclass the Context class and make `convert()` a method of the context itself.

Reply:
Finally getting some time to work on this.  I had pulled the base as a flag in the context and then was thinking of using `setBase` to do this as well (and will need to do something similar for digits), but it seems that `setBase` has no access to the context.   My latest try is to call setBase like:

```
context::NondecimalBase->setBase(5);
```
and then in the `setBase` function, I try to get the context via:

```
sub setBase {
	my ($name, $base) = @_;
	my $context = $main::context{$name} = Parser::Context->getCopy($name);
}
```

but this is undefined.  I was looking at `contextAlternativeDecimal.pl` as an example.